### PR TITLE
Add validation and rotation of certificate for clusterinterceptor

### PR DESCRIPTION
**Pull request by savitaashture**
# Changes
* Added validation to verify certs validity
Fixes https://github.com/tektoncd/triggers/issues/1354#issuecomment-1170104987
* Added changes to rotate certificates if expired

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#tests) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#docs) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commits)
- [ ] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/triggers/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

```release-note
Triggers Interceptor now does validation and rotation of certificates if expires 
```

### CODE REVIEW

**Review by khrm — COMMENTED** [review]
/test pull-tekton-triggers-integration-tests
/approve

File: cmd/interceptors/main.go [hunk 1]
```diff
@@ -217,3 +221,46 @@ func getCerts(ctx context.Context, secretLister v1.SecretLister, kubeClient *kub
 	}
 	return keyFile, certFile, caCert, nil
 }
+
+func checkCertValidity(serverCert, caCert []byte, logger *zap.SugaredLogger) {
```
[the remark this answers is not in the document]
Comment (savitaashture) [reply cmd/interceptors/main.go]: Updated 
PTAL
Thank you

File: pkg/interceptors/server/server.go [hunk 2]
```diff
@@ -236,3 +240,46 @@ func (is *Server) UpdateCRDWithCaCert(ctx context.Context, triggersV1Alpha1 trig
...
+						logger.Errorf("failed to parse certificate: %v", err.Error())
+					}
+				}
+
+				opts := x509.VerifyOptions{
+					Roots: roots,
+				}
+
+				if _, err := cert.Verify(opts); err != nil {
+					logger.Errorf("failed to verify certificate: %v", err.Error())
```
Comment (dibyom) [inline pkg/interceptors/server/server.go:276]: So, if we fail to validate the certificate - what do we do next? It seems like all we are doing is logging an error, should this be an indication to update the cert?

File: pkg/interceptors/server/server_test.go [hunk 3]
```diff
@@ -274,3 +274,15 @@ func Test_UpdateCRDWithCaCert(t *testing.T) {
 		t.Error(err)
 	}
 }
+
+func Test_CheckCertValidity(t *testing.T) {
+	// Adding delay of a minute so that ticker will be triggered
+	time.Sleep(time.Minute)
```
Comment (dibyom) [inline pkg/interceptors/server/server_test.go:280]: I think if we could pass in the ticker via a parameter, we could better test just the checkCertValidity function

Comment (dibyom) [reply pkg/interceptors/server/server_test.go:280]: Another option would be to pass in a [clock.Ticker](https://pkg.go.dev/k8s.io/utils@v0.0.0-20220728103510-ee6ede2d64ed/clock#Ticker) to the function above and then use https://pkg.go.dev/k8s.io/utils/clock/testing#FakeClock.NewTicker in the test

File: pkg/interceptors/server/server.go (hunk 2 above)
Comment (savitaashture) [reply pkg/interceptors/server/server.go:276]: Yes its an indication to update the cert

**Comment by savitaashture** [comment]
@dibyom @khrm 

PR is ready for review 
PTAL :pray:

File: cmd/interceptors/main.go [hunk 4]
```diff
@@ -82,26 +79,41 @@ func main() {
...
 
-	if err := listAndUpdateClusterInterceptorCRD(ctx, tc, service, caCert); err != nil {
+	if err := service.ListAndUpdateClusterInterceptorCRD(ctx, tc.TriggersV1alpha1(), caCert); err != nil {
 		return
 	}
+
+	// After creating certificates using CreateCerts lets validate validity of created certificates
+	service.CheckCertValidity(ctx, serverCert, caCert, kubeclient.Get(ctx).CoreV1(), logger, tc.TriggersV1alpha1(), time.Minute)
+
+	interceptorSecretName := os.Getenv("INTERCEPTOR_TLS_SECRET_NAME")
```
Comment (khrm) [inline cmd/interceptors/main.go:99]: `INTERCEPTOR_TLS_SECRET_NAME` should be under some constant.

Comment (savitaashture) [reply cmd/interceptors/main.go]: Done

File: cmd/interceptors/main.go [hunk 5]
```diff
@@ -82,26 +79,41 @@ func main() {
 	mux.Handle("/", service)
 	mux.HandleFunc("/ready", handler)
 
-	keyFile, certFile, caCert, err := server.CreateCerts(ctx, kubeClient.CoreV1(), logger)
+	tc, err := triggersclientset.NewForConfig(cfg)
```
[the remark this answers is not in the document]
Comment (savitaashture) [reply cmd/interceptors/main.go:82]: Sure will do in subsequent PR

[the remark this answers is not in the document]
Comment (savitaashture) [reply cmd/interceptors/main.go:82]: Created [issue](https://github.com/tektoncd/triggers/issues/1441)  to track

**Pull request merged**